aggregator/src/aggregator/garbage_collector.rs

```diff
@@ -0,0 +1,240 @@
+use std::sync::Arc;
+
+use crate::{
+    datastore::Datastore,
+    messages::TimeExt,
+    task::{QueryType, Task},
+};
+use anyhow::{anyhow, Context, Result};
+use futures::future::join_all;
+use janus_core::time::Clock;
+use janus_messages::Role;
+use tracing::error;
+
+pub struct GarbageCollector<C: Clock> {
+    // Dependencies.
+    datastore: Arc<Datastore<C>>,
+    clock: C,
+}
+
+impl<C: Clock> GarbageCollector<C> {
+    pub fn new(datastore: Arc<Datastore<C>>, clock: C) -> Self {
+        Self { datastore, clock }
+    }
+
+    #[tracing::instrument(skip(self))]
+    pub async fn run(&self) -> Result<()> {
+        // TODO(#224): add support for handling only a subset of tasks in a single job (i.e. sharding).
+
+        // Retrieve tasks.
+        let tasks = self
+            .datastore
+            .run_tx(|tx| Box::pin(async move { tx.get_tasks().await }))
+            .await;
+        let tasks = match tasks {
+            Ok(tasks) => tasks,
+            Err(error) => return Err(error).context("couldn't retrieve tasks"),
+        };
+
+        // Run GC for each task.
+        join_all(tasks.into_iter().map(|task| async move {
+            let task = Arc::new(task);
+            if let Err(err) = self.gc_task(Arc::clone(&task)).await {
+                error!(task_id = ?task.id(), ?err, "Couldn't GC task");
+            }
+        }))
+        .await;
+        Ok(())
+    }
+
+    async fn gc_task(&self, task: Arc<Task>) -> Result<()> {
+        let oldest_allowed_report_timestamp =
+            if let Some(report_expiry_age) = task.report_expiry_age() {
+                if task.role() != &Role::Leader || task.query_type() != &QueryType::TimeInterval {
+                    return Err(anyhow!(
+                        "garbage collection is implemented only for leader, time-interval tasks"
```

Comment:
I get why we can't support fixed size tasks in the 0.2 branch, but is there some reason GC is impossible to implement for the helper or is that a follow up task?

Reply:
Yes: helper GC would imply cleaning up the `aggregation_jobs` table for helper tasks. Doing so requires knowing which aggregation jobs are covered by a `collect_jobs` row. But this isn't easy to determine, since (a) a helper `aggregation_jobs` row has no `batch_interval`/`batch_identifier`, and (b) a leader is free to create an aggregation job containing reports spanning an arbitrary interval of time, not necessarily one that fits into a given collect job interval. (Janus won't create such aggregation jobs.)